Issue #6068: migrate to picocli command line parser from Commons CLI

remkop opened the pull request:
Issue #6068: migrate to picocli command line parser from Commons CLI

This PR replaces #6082.

Sorry it took so long. It was harder than I thought to get all the tests and checks to pass. :-)

remkop commented:
I may need some help with the failing CI.
The IDEA Inspections Pull Request (Checkstyle) — TeamCity build failed with

```
[12:13:00]Number of inspection errors 16 is 16 more than the provided threshold 0
[12:13:00]Build failure condition: Fail build if number of inspection errors is more than 0
[12:13:00]Number of inspection warnings 22 is 22 more than the provided threshold 0
[12:13:00]Build failure condition: Fail build if number of inspection warnings is more than 0
```

ci/circleci: pitest2 — tests failed on CircleCI:

(at first glance seems a problem in PowerMockJUnit44RunnerDelegateImpl)

```
12:10:10 PM PIT >> INFO : Created  1 mutation test units
stderr  : 12:11:48 PM PIT >> SEVERE : Error while running adapter JUnit fixture class com.puppycrawl.tools.checkstyle.MainTest with filter Optional[testTreeWalkerThreadsNumber(com.puppycrawl.tools.checkstyle.MainTest)]
java.lang.RuntimeException: PowerMock internal stderr  : error: Should never throw exception at this level
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.handleException(PowerMockJUnit44RunnerDelegateImpl.java:384)
	at org.powermock.modules.junit4.istderr  : nternal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.evaluateStatement(PowerMockJUnit47RunnerDelegateImpl.java:110)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.execustderr  : teTest(PowerMockJUnit47RunnerDelegateImpl.java:82)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.runBeforesThenTestThenAfters(PowerMockJUnit44RunnerDelegateImpl.java:298)
	at org.junit.internstderr  : al.runners.MethodRoadie.runTest(MethodRoadie.java:87)
	at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:50)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.invokeTestMethod(PowerMockJUnit44RunnerDelegateImpstderr  : l.java:218)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.runMethods(PowerMockJUnit44RunnerDelegateImpl.java:160)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$1.run(PowerMockJUnit44Rustderr  : nnerDelegateImpl.java:134)
	at org.junit.internal.runners.ClassRoadie.runUnprotected(ClassRoadie.java:34)
	at org.junit.internal.runners.ClassRoadie.runProtected(ClassRoadie.java:44)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelestderr  : gateImpl.run(PowerMockJUnit44RunnerDelegateImpl.java:136)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.run(JUnit4TestSuiteChunkerImpl.java:117)
	at org.powermock.modules.junit4.common.internal.impl.AbstractCommonPowerMocstderr  : kRunner.run(AbstractCommonPowerMockRunner.java:57)
	at org.powermock.modules.junit4.PowerMockRunner.run(PowerMockRunner.java:59)
	at org.pitest.junit.adapter.CustomRunnerExecutor.run(CustomRunnerExecutor.java:42)
	at org.pitest.junit.adapter.AdaptedJUnitTestderr  : stUnit.execute(AdaptedJUnitTestUnit.java:69)
	at org.pitest.mutationtest.execute.MutationTimeoutDecorator.lambda$createRunnable$0(MutationTimeoutDecorator.java:83)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.costderr  : ncurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.hanstderr  : dleException(PowerMockJUnit44RunnerDelegateImpl.java:362)
	... 20 more
```

rnveach commented:
> IDEA Inspections Pull Request 

You have to view the report at https://teamcity.jetbrains.com/viewLog.html?buildId=1628574&tab=Inspection&buildTypeId=Checkstyle_IdeaInspectionsPullRequest .
When you first enter the CI and sign in for anonymous access, there is a link that says view report.

> pitest2 

I am not sure if that exception is related. Usually the CI will tell you what is missing, but it looks like it failed to do that for some reason.
@romani Please look.

> Mutation score of 99 is below threshold of 100

You may have to run pitest locally and look at the report to see what is missing.

remkop commented:
@rnveach, @romani, Please consider this PR. Sorry for the delay. I found it quite difficult to get the CI checks to pass (I’m relieved they pass now; I was close to giving up a few times). :-)

This PR migrates the `Main` and `JavadocPropertiesGenerator` command line utilities from Commons CLI to picocli. 

Note that I used picocli’s [mixinStandardHelpOptions](https://picocli.info/#_mixin_standard_help_options) function to provide `--help` and `--version` options without the need for an annotated field. I hope you agree that this is useful. I updated the documentation accordingly. 

@rnveach, I considered moving the annotated fields into an inner class but I found the current code to be cleaner: we would end up either moving a significant portion of the logic into the inner class, or alternatively referencing the option fields indirectly like this:

```java
if (options.printAst || options.printAstWithComments
        || options.printJavadocTree || options.printTreeWithJavadoc) {
    if (options.suppressionLineColumnNumber != null 
            || options.configurationFile != null 
            || options.propertiesFile != null 
            || options.outputPath != null
...
```

The current code (without the `options` reference) seems simpler and cleaner. I hope you agree.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -373,92 +359,60 @@ else if (cmdLine.hasOption(OPTION_C_NAME)) {
-                        return record.getLoggerName().startsWith(packageName);
-                    }
-                });
+                handler.setFilter(new OnlyCheckstyleLoggersFilter());
```
What was the matter with the original way the filter was done?

remkop replied:
One of the checks (I forgot which one) failed with "anonymous inner class can be replaced with static inner class".

remkop commented on pom.xml:
```diff
@@ -956,6 +958,46 @@
+                    <limit>
+                      <counter>LINE</counter>
+                      <value>COVEREDRATIO</value>
+                      <minimum>0.98</minimum>
```
Yes, I could not get it to cover the System.exit call.
The previous version had a test that mocked the System.exit static method but this test always failed for me.

remkop commented on config/intellij-idea-inspections.xml:
```diff
@@ -2677,10 +2677,7 @@
-                     enabled_by_default="true">
-        <option name="queries" value="trace,write,forEach"/>
-        <option name="updates" value="pop,push,shift,splice,unshift"/>
-    </inspection_tool>
```
This was unintentional and should not have been committed.

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -259,111 +261,95 @@ public static void main(String... args) throws IOException {
-        final CommandLineParser clp = new DefaultParser();
-        // always returns not null value
-        return clp.parse(buildOptions(), args);
+    @Override public Integer call() throws IOException, CheckstyleException {
```
ok will do

remkop commented on config/spotbugs-exclude.xml:
```diff
@@ -107,6 +107,46 @@
         <Bug pattern="NP_NULL_PARAM_DEREF"/>
     </Match>
+    <Match>
+        <!-- false-positive: field values are injected by picocli -->
```
Thank you for the tip! Will do.

remkop replied:
Fixed. (Commit amended with force-push to keep history clean. Let me know if you prefer additional commits instead of amending commits.)

remkop commented:
@rnveach Fixed some of the things you pointed out. I amended the commit with force-push to keep history clean. Is that the preferred way or do you prefer additional commits in response to feedback?

remkop commented:
@rnveach I made a change was able to get back to 100% coverage.

remkop commented:
@rnveach , @romani  I believe I have made all the requested changes and hope this is ready to be merged. Could you take another look?

remkop commented on config/intellij-idea-inspections.xml:
```diff
@@ -2945,7 +2945,10 @@
          for this purpose -->
     <inspection_tool class="LocalCanBeFinal" enabled="false" level="ERROR"
-                     enabled_by_default="false"/>
+                     enabled_by_default="false">
```
Good point. I will revert this change.

remkop replied:
Done

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/JavadocPropertiesGenerator.java:
```diff
@@ -78,29 +85,27 @@ private JavadocPropertiesGenerator() {
-            printUsage();
+        catch (ParameterException ex) {
+            System.err.println(ex.getMessage());
+            ex.getCommandLine().usage(System.err);
```
The reason is that error reporting for invalid user input is much nicer: instead of a stack trace, the user will see an error message indicating what they did wrong and the usage help message showing what options are available.
You are correct that this changes the behaviour but I hope you agree this is slightly more user-friendly.

For errors other than invalid user input, the exception is passed up the chain as before.

remkop commented:
@rnveach Thank you for the additional feedback! I've removed the redundant exclusion from intellij-idea-inspections.xml.

@romani It would be great if you could take a look to see if this is ready to be merged.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -259,111 +261,96 @@ public static void main(String... args) throws IOException {
-        // always returns not null value
-        return clp.parse(buildOptions(), args);
+    @Override
+    public Integer call() throws IOException, CheckstyleException {
```
@romani This is an overridden method, so it can't be renamed. The class implements `java.util.concurrent.Callable` which is forcing the `call` method name.

`Integer` is because `Callable` has a template and I assume it accept primitive types.

remkop replied:
Will do.

remkop replied:
Done

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -53,12 +49,27 @@
+        + "reported to standard out in plain format. Checkstyle requires a configuration "
+        + "XML file that configures the checks to apply.",
+    mixinStandardHelpOptions = true, versionProvider = Main.VersionProvider.class)
+public final class Main implements Callable<Integer> {
```
I don't know the full details, but it seems [`picocli.CommandLine` only works with Runnable/Callable](https://github.com/remkop/picocli/blob/d27d7a8dd6885d1cc0672d983c45bb84329e166d/src/main/java/picocli/CommandLine.java#L829) and we are passing it our `Main`.

remkop replied:
I will remove the `implements Callable`, rename `call` to `execute` and move the options into an inner class as requested.

@rnveach @romani  To clarify, picocli does not _require_ client code to be `Runnable` or `Callable`. It's just that _if_ your application implements one of these interfaces, you can choose to use one of picocli's convenience methods to run your application in one line of code:

```
CommandLine.call(new Main(), args);
```

This one single line takes care of parsing the arguments, validating user input, checking whether the user requested version/usage help, and finally executing the business logic in the `call` method.

This is what I initially tried to achieve. 

However, after taking into account the error codes and exception handling, it became a lot less clean and in the end I just used the `parseArgs` method. In that case, `implements Callable` is no longer necessary as you pointed out.

remkop replied:
Done (no longer implements Callable)

rnveach commented on .ci/jsoref-spellchecker/whitelist.words:
```diff
@@ -317,6 +317,7 @@ Dfoo
 Dforbiddenapis
 Dgpg
 Dgui
+dhj
```
[See `MainTest`](https://github.com/checkstyle/checkstyle/pull/6126/files#diff-9cb7b23a8bfd6d92a3a35bdba7bac261R73).
> `"Usage: checkstyle [-dhjJtTV] [--executeIgnoredModules] [-gxs] [-c=<configuration"`

remkop commented on src/xdocs/cmdline.xml.vm:
```diff
@@ -143,7 +143,10 @@ java -D&lt;property&gt;=&lt;value&gt;  \
         </li>
         <li>
-          <code>-v</code> - print product version and exit. Any other option is ignored.
+          <code>-V, --version</code> - print product version and exit. Any other option is ignored.
```
Most unix tools use `-V` (uppercase) for `--version` and `-v` (lowercase) for `--verbose`. This is not a standard but fairly common usage, as documented [here](http://catb.org/~esr/writings/taoup/html/ch10s05.html#id2948149).  Example tools that follow this convention: grep, curl, gzip, ssh.

Because of this background, picocli has a shortcut for providing `-h`, `--help` for usage help, and `-V`, `--version` for version help: these options are "mixed in" when the command is defined as `@Command(mixinStandardHelpOptions = true)`. I used this shortcut to avoid the need for defining annotated fields for these standard options.

Unfortunately the shortcut does not allow customizing the option names. If the name needs to be changed, we need to define an annotated field: `@Option(names = "-v") boolean isVersionRequested;`. I thought it better to use the `mixinStandardHelpOptions` shortcut.

romani replied:
good to know.
We can "-V" I see no problem with it.

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -194,47 +217,29 @@ private Main() {
-                    errorCounter = runCli(commandLine, filesToProcess);
-                    exitStatus = errorCounter;
+            final ParseResult parseResult = commandLine.parseArgs(args);
+            final boolean helpWasRequested = CommandLine.printHelpIfRequested(parseResult);
```
Will do.

remkop replied:
I will change this to use the `ParseResult::isUsageHelpRequested` and `ParseResult::isVersionHelpRequested` methods.

remkop replied:
Done

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -194,47 +217,29 @@ private Main() {
-        boolean cliViolations = false;
+
+        final Main main = new Main();
+        final CommandLine commandLine = new CommandLine(main);
```
Will do.

remkop replied:
An instance is needed because the annotated fields are injected with the command line option values. This can be an inner class. I will make the change.

rnveach replied:
@remkop For your util, you pass in the instance. When would your `CommandLine` work with a non-default constructor instance? To me, it would seem the majority of main classes would use the default constructor.
You could have `CommandLine` accept `Class<?>` as well and just call `clss.newInstance()`.

remkop replied:
When parsing the command line, picocli will convert the command line options to the type of the annotated fields and inject the values into the fields. This is one of the advantages of using picocli vs. Commons CLI: usually you don’t need to inspect a `ParseResult`, you just use your annotated “bean” directly. 

So, we need the instance to get the parse results. For example:

```
class CliOptions {
  @Option(names = “-x”)
  int x;
}

CliOptions options = new CliOptions();
CommandLine cmd = new CommandLine(options);
String[] args = { “-x”, “123” };
cmd.parseArgs(args);
assert options.x == 123; // use instance
```

This is a common usage pattern. There are variations where you don’t need an instance, for example if the class implements `Runnable` or `Callable`. (And to support such use cases it is already possible to pass a `Class` to the `CommandLine` constructor and picocli will internally create an instance just like you describe.)

However, this application has requirements for error handling that means it’s simpler to just use the “common usage pattern” instead of the `Runnable` / `Callable` convenience methods.

remkop replied:
Done

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -75,114 +86,126 @@
+
+    /** Allows access to the command line parser model. */
+    @Spec
+    private CommandSpec commandSpec;
```
Will do.

remkop replied:
Done (field removed)

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -191,50 +113,35 @@ private Main() {
+        catch (ParameterException ex) {
+            System.err.println(ex.getMessage());
+            ex.getCommandLine().usage(System.err);
+            exitStatus = EXIT_WITH_INVALID_USER_INPUT_CODE;
```
please be consistent with next catch:
```
<set exit codes ...>
<print messages>
```

remkop replied:
Done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -259,613 +163,588 @@ public static void main(String... args) throws IOException {
+         * @throws IOException if a file could not be read.
+         * @throws CheckstyleException if something happens processing the files.
+         */
+        private int runCli(List<File> filesToProcess) throws IOException, CheckstyleException {
```
Why CliOption now it in change to run Checkstyle ?
Cli is just one way to configure execution, and it collects all parameters, and this class should only do CLI related logic and nothing else. All orchestration of how to run checkstyle should stay in Main class.

remkop replied:
I see your point that it is better to focus on the responsibilities of each class.

I just submitted a new PR (now going through CI) where responsibilities are divided like this:

**CliOptions responsibilities**

* defines the command line options and usage help message descriptions
* validating the user input
* listing the target files (after figuring out regex patterns, exclusions, etc.)

**Main responsibilities** 

* delegate some work to CliOptions
* run checkstyle


![checkstyle](https://user-images.githubusercontent.com/5917763/47273096-d3999280-d5c9-11e8-8c36-466cf35160b2.png)

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/JavadocPropertiesGenerator.java:
```diff
@@ -78,29 +85,27 @@ private JavadocPropertiesGenerator() {
-            writePropertiesFile(inputFile, outputFile);
+    public static void main(String... args) throws CheckstyleException {
+        final JavadocPropertiesGenerator generator = new JavadocPropertiesGenerator();
+        final CommandLine cmd = new CommandLine(generator).setUsageHelpWidth(USAGE_HELP_WIDTH);
```
Will do.

remkop replied:
Done

checkstyle-jenkins commented:
Build succeeded

romani commented on config/pmd-main.xml:
```diff
@@ -19,5 +19,21 @@
+      the picocli and not by the application.
+      They may appear unused by static code analysis tools. -->
+    <rule ref="category/java/bestpractices.xml/UnusedPrivateField">
+        <property name="ignoredAnnotations"
```
@remkop , please keep replying to each point as "done", it helps.

remkop replied:
Done (properties tag added)

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -717,155 +439,319 @@ private static OutputStream getOutputStream(String outputLocation) throws IOExce
+         * Determines the files to process.
+         * @return list of files to process
+         */
+        private List<File> getFilesToProcess() {
```
can this , and few method below be returned to Main ?

romani replied:
@remkop ,
I might miss smth , but , it is use one place at https://github.com/checkstyle/checkstyle/blob/ad9d8ec298752a589597c3aae58a30da253fad48/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L169 that in Main class. 
It will require cli.getExclusions() but we can change method and put report of this method as argument.

remkop replied:
Ok, let me take another look.

remkop commented on config/pmd-main.xml:
```diff
@@ -19,5 +19,16 @@
     </rule>
-
+    <rule ref="category/java/design.xml/ImmutableField">
+        <property name="ignoredAnnotations"
```
Done

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -75,114 +75,36 @@
-    /** A string value of 1. */
-    private static final String ONE_STRING_VALUE = "1";
+    /** Default output format. */
+    private static final OutputFormat DEFAULT_OUTPUT_FORMAT = OutputFormat.PLAIN;
```
Done

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -75,114 +75,36 @@
-    /** Default distance between tab stops. */
-    private static final String DEFAULT_TAB_WIDTH = "8";
+    /** Option name for output format. */
+    private static final String OUTPUT_FORMAT_OPTION = "-f";
```
Done

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -75,114 +75,36 @@
-    /** Name for 'plain' format. */
-    private static final String PLAIN_FORMAT_NAME = "plain";
+    /** Default distance between tab stops. */
+    private static final int DEFAULT_TAB_WIDTH = 8;
```
Done

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -75,114 +75,36 @@
-    /** Name for the option '--treeWithJavadoc'. */
-    private static final String OPTION_TREE_JAVADOC_NAME = "treeWithJavadoc";
+    /** The default number of threads to use for checker and the tree walker. */
+    private static final int DEFAULT_THREAD_COUNT = 1;
```
Done

remkop commented on config/spotbugs-exclude.xml:
```diff
@@ -107,6 +107,20 @@
         <Bug pattern="NP_NULL_PARAM_DEREF"/>
     </Match>
+    <Match>
+        <!-- false-positive: field values are injected by picocli -->
```
Done

remkop commented on config/pmd.xml:
```diff
@@ -36,7 +36,10 @@
-      <property name="violationSuppressXPath" value="//ClassOrInterfaceDeclaration[@Image='Main']"/>
+      <property name="violationSuppressXPath"
+                value="//ClassOrInterfaceDeclaration
+                       [@Image='Main' or @Image='Main$CliOptions' or @Image='CliOptions'
```
Done

remkop commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -75,114 +86,126 @@
+    private int treeWalkerThreadsNumber = DEFAULT_THREAD_COUNT;
+
+    /** The files to process (the specified source files without the exclusions). */
+    private List<File> filesToProcess;
```
Done (field removed)

remkop commented:
@romani Thanks for the detailed review. All items you mentioned have been addressed. CI build is green. Please take another look. 

Out of interest, what time zone are you in? I’m in Tokyo (JST).

romani commented:
I am in PST

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -75,114 +75,24 @@
     public static final String CREATE_LISTENER_EXCEPTION = "Main.createListener";
+
+    /** Name for the moduleConfig attribute 'tabWidth'. */
+    private static final String ATTRIB_TAB_WIDTH_NAME = "tabWidth";
```
please move this constant to CliOptions class, to let this class own all about cli.

remkop replied:
Done: moved ATTRIB_TAB_WIDTH_NAME to CliOptions.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -75,114 +75,24 @@
     private static final Log LOG = LogFactory.getLog(Main.class);
 
     /** Width of CLI help option. */
     private static final int HELP_WIDTH = 100;
```
please move this constant to CliOptions class, to let this class own all about cli.

remkop replied:
Done: Moved HELP_WIDTH to CliOptions also.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Main.java:
```diff
@@ -259,206 +151,85 @@ public static void main(String... args) throws IOException {
-                "TreeWalker threads number must be greater than zero",
-                "Invalid TreeWalker threads number");
+        // return error if something is wrong in arguments
+        final List<File> filesToProcess = getFilesToProcess(options);
```
getFilesToProcess need to be moved right after `execute`, to let read code as text by simply scrolling from to bottom.
the same problem is with getVersionString, just need to be moved to be right after `main`.



please do not be shocked, but we do really care about this :)
One day it will be completely automated, for now we did only half of way to this target.

Here is diagrams to help see it visually:
https://methods-distance.herokuapp.com/dsm?source_url=https://raw.githubusercontent.com/checkstyle/checkstyle/ea98b563e16b0c2bc09ee61d2824bdf0f1ea4a86/src/main/java/com/puppycrawl/tools/checkstyle/JavadocPropertiesGenerator.java
https://methods-distance.herokuapp.com/dsm?source_url=https://raw.githubusercontent.com/checkstyle/checkstyle/ea98b563e16b0c2bc09ee61d2824bdf0f1ea4a86/src/main/java/com/puppycrawl/tools/checkstyle/Main.java

first link - all is fine.
second link show a problem, minor problem.

Sorry that we are so fanatic about easy to read code.

remkop replied:
Done:

* moved `getFilesToProcess` to after `execute`
* moved `getVersionString` to after `main`

romani commented:
@rnveach , please start code review.

Here is how it works:
```
$ java -jar checkstyle-8.15-SNAPSHOT-all.jar
Missing required parameter: <files>
Usage: checkstyle [-dhjJtTV] [--executeIgnoredModules] [-gxs] [-c=<configurationFile>]
                  [-C=<checkerThreadsNumber>] [-f=<format>] [-o=<outputPath>] [-p=<propertiesFile>]
                  [-s=<suppressionLineColumnNumber>] [-tabWidth=<tabWidth>]
                  [-W=<treeWalkerThreadsNumber>] [-e=<exclude>]... [-x=<excludeRegex>]... <files>...
Checkstyle verifies that the specified source code files adhere to the specified rules. By default
errors are reported to standard out in plain format. Checkstyle requires a configuration XML file
that configures the checks to apply.
      <files>...             One or more source files to verify
      --executeIgnoredModules
                             Allows ignored modules to be run.
  -c= <configurationFile>    Sets the check configuration file to use.
  -C, --checker-threads-number=<checkerThreadsNumber>
                             (experimental) The number of Checker threads (must be greater than zero)
  -d, --debug                Print all debug logging of CheckStyle utility
  -e, --exclude=<exclude>    Directory path to exclude from CheckStyle
  -f= <format>               Sets the output format. Valid values: xml, plain. Defaults to plain
      -gxs, --generate-xpath-suppression
                             Generates to output a suppression.xml to use to suppress all violations
                               from user's config
  -h, --help                 Show this help message and exit.
  -j, --javadocTree          Print Parse tree of the Javadoc comment
  -J, --treeWithJavadoc      Print full Abstract Syntax Tree of the file
  -o= <outputPath>           Sets the output file. Defaults to stdout
  -p= <propertiesFile>       Loads the properties file
  -s= <suppressionLineColumnNumber>
                             Print xpath suppressions at the file's line and column position. Argument
                               is the line and column number (separated by a : ) in the file that the
                               suppression should be generated for
  -t, --tree                 Print Abstract Syntax Tree(AST) of the file
  -T, --treeWithComments     Print Abstract Syntax Tree(AST) of the file including comments
      -tabWidth=<tabWidth>   Sets the length of the tab character. Used only with "-s" option. Default
                               value is 8
  -V, --version              Print version information and exit.
  -W, --tree-walker-threads-number=<treeWalkerThreadsNumber>
                             (experimental) The number of TreeWalker threads (must be greater than zero)
  -x, --exclude-regexp=<excludeRegex>
                             Regular expression of directory to exclude from CheckStyle
```

remkop commented:
1. Yes, for a long usage help message like CheckStyle's it may make sense to just show an error message  after the user specified invalid input and inform the user they can get detailed help with `checkstyle --help`. Future versions of picocli will [do this automatically](https://github.com/remkop/picocli/issues/480) for applications that use the `CommandLine::run` and `CommandLine::call`  convenience methods.
3. I changed `-tabWidth` to `--tabWidth` with two leading hyphens
4. I agree that having `-g` looks better in the usage help. I added `-g`, and kept `-gxs` to avoid breaking existing 3rd party tools that are using this option. Let me know if you want to remove `-gxs` altogether.
5. This is a display issue in the picocli usage help message layout for short options. I raised https://github.com/remkop/picocli/issues/531 for this.

remkop commented:
All done:

*  `-gxs` removed
* Checkstyle now prints the error message followed by a short usage message and a hint to use `checkstyle --help` for more information after the user specified invalid input. The full usage help message is only printed on user request.
* I created issue https://github.com/checkstyle/checkstyle/issues/6175 as a reminder for the display issue in the usage help (the spacing issue). Glad this is not a showstopper. :-)
* I moved  listFiles to after getFilesToProcess, and also some other methods so they can be read more fluently

remkop commented:
I had to resubmit a few times to get past transient CI build failures, but all checks passed now.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:
```diff
@@ -65,45 +68,62 @@
+            + "Try 'checkstyle --help' for more information.%n");
+
     private static final String USAGE = String.format(Locale.ROOT,
-          "usage: java com.puppycrawl.tools.checkstyle.Main [options] -c <config.xml>"
```
@romani Do we need to create a section documenting how some arguments changed between the versions?
See my next comment on why I am asking.

@remkop Some other applications I have created use the CLI as an api to execute checkstyle.
I know some parameters changed, but I don't know the extent. If any of these are affected, please let me know.
https://github.com/checkstyle/contribution/blob/master/checkstyle-web-tester/checkstyle.php#L136
https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff.sh#L208
https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff_antlr.sh#L180
https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/sevntu_launch_diff.sh#L208

romani replied:
> Do we need to create a section documenting how some arguments changed between the versions?

we can make comment in issue, to reference to it if somebody ask a questions. As far as know we change only "-gxs" (removed) and "-V" for version, all other stay the same. Minor change.

remkop replied:
I added this comment that summarizes the option name changes: https://github.com/checkstyle/checkstyle/issues/6068#issuecomment-434097233

remkop replied:
@rnveach I checked the 4 links above and I believe none of them should be impacted by the changes in this PR.

rnveach replied:
@remkop Thanks.

romani commented:
> Yes, for a long usage help message like CheckStyle's it may make sense to just show an error message after the user specified invalid input and inform the user they can get detailed help with checkstyle --help. Future versions of picocli will do this automatically for applications that use the CommandLine::run and CommandLine::call convenience methods.

This is still an issue on our side, we are not going to use such convenience methods, they are against our design ideas. I do believe we will not be alone to about extra "extends" and do more compositions.
How picocli can support our use case ?

remkop commented:
@romani  I think there is a misunderstanding: in my last update for the PR I already implemented your request: Checkstyle now prints the error message followed by a short usage message and a hint to use `checkstyle --help` for more information after the user specified invalid input. The full usage help message is only printed on user request.

I believe this is exactly what you asked for. Did I misunderstand?

Please disregard my earlier comment on my plans for supporting this in future versions of picocli. Since Checkstyle does not use the convenience methods, it will not be impacted by such future changes.

remkop commented on src/xdocs/cmdline.xml.vm:
```diff
@@ -40,8 +40,8 @@ java -D&lt;property&gt;=&lt;value&gt;  \
      -c &lt;configurationFile&gt; \
      [-f &lt;format&gt;] [-p &lt;propertiesFile&gt;] [-o &lt;file&gt;] \
      [-s &lt;line:column&gt;] [-gxs | --generate-xpath-suppression] [-tabWidth &lt;length&gt;] \
```
Good find. Both done:

* `-gxs` replaced by `-g` in synopsis as well as the option description below
* verified other options and made two updates to the documentation:
    * `-tabWidth` became `--tabWidth` as agreed 
    *  `-executeIgnoredModules` became `--executeIgnoredModules` as agreed

remkop commented on config/pmd-main.xml:
```diff
@@ -19,5 +19,25 @@
+    </rule>
+    <!-- Some fields with picocli annotations will be only used by
+      the picocli and not by the application.
+      They may appear unused by static code analysis tools. -->
```
Done. I updated the PR. Waiting for the CI pipeline to pass now.

romani commented:
> I believe this is exactly what you asked for. Did I misunderstand?

my bad, I did not noticed this change. Here is how it works in comparison with "grep":
```
$ java -jar checkstyle-8.15-SNAPSHOT-all.jar -c 1.xml -tabWidth 3 ../pom.xml 
Unknown option: -abWidth
Usage: checkstyle [OPTIONS]... FILES...
Try 'checkstyle --help' for more information.

$ grep -TTT
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
```

looks good.
I am done with review. PR is approved.
@remkop , thanks a lot for your hard work.

remkop commented:
I’m hoping that other people will follow CheckStyle’s example and adopt picocli in their project. 

If you like what you’ve seen so far, please say some nice things about picocli in a few public places. It really helps!

romani commented:
I will twit about it, ones we merge your PR.

Travis was restarted, it was failed for unrelated issues.

Teamcity is complaining due to time out (30 min is limit). Here is history of builds:
```
27 Oct 18 05:22 25m:45s
26 Oct 18 04:23 23m:55s
26 Oct 18 00:41 25m:31s
23 Oct 18 20:38 24m:00s
21 Oct 18 18:14 26m:43s
20 Oct 18 22:24 24m:00s
20 Oct 18 20:47 25m:29s
20 Oct 18 20:18 14m:28s
20 Oct 18 19:58 26m:25s
20 Oct 18 17:41 27m:46s
20 Oct 18 17:23 26m:47s
16 Oct 18 13:53 30m:07s
16 Oct 18 13:31 26m:36s
16 Oct 18 00:53 24m:49s
14 Oct 18 15:28 25m:11s
7 Oct 18 18:09 21m:48s
6 Oct 18 15:02 21m:24s
4 Oct 18 14:50 19m:16s
```
